### PR TITLE
redeploy banner by region

### PR DESCRIPTION
**Pull request by tsop14**
## What does this change?
Adds the ability to redeploy the engagement banner (big yellow one) in different 'reader revenue regions'. 

## Screenshots
screen shot 2018-12-13 at 15 45 43

## What is the value of this and can you measure success?
Non-engineers able to redeploy to different regions. Will be even more useful once non-engineers can create region specific tests.

## Changes

Clientside
- Creates `readerRevenueRegion` type. Currently matches the areas we are running campaigns in: gb, us, au, row. Note in tests etc. we refer to `UK` but as this uses geolocation, it refers to `gb`. Also note that these are lower case but geolocation is stored in upper case. 
- Checks user's geolocation and maps to a `readerRevenueRegion`.
- Updates bannerLastClosed in local storage to hold an objects with timestamps by region, i.e.
`value: {gb: "2018-12-13T15:46:18.149Z", row: "2018-12-13T16:02:12.972Z", us: "2018-12-13T16:03:50.834Z"}`

- When the user closes the banner, it overwrites the relevant region timestamp. 
- When deciding if the user should see the banner, uses their geolocation to determine their `readerRevenueRegion` and requests timestamp from relevant apps endpoint (see details below).
- Note, as before, if there is nothing set in local storage the user sees the banner. If fetch for timestamp fails, the user does not see the banner.

Serverside
- Creates ReaderRevenueRegion types. 
- Creates endpoint for each region: read from s3 bucket e.g. `/reader-revenue/contributions-banner-deploy-log/uk`, update with `/reader-revenue/redeploy-contributions-banner/gb`.  Could have had one endpoint and read and updated the s3 file but felt simpler to continue overwriting the timestamp and create one per region. The endpoint is cached for 7 days.
- Adds a button to reader revenue section in admin to update the timestamp for each region. As before, message with the user who redeployed the button is written to logs and this information is available for 7 days. 

> What happens if I want to redeploy all at the same time?

Click, click, click, click. Death by buttons!!! 

### Tested

- [x] Locally  note, when testing locally, you need to comment out [cache purging](https://github.com/guardian/frontend/compare/em-deploy-banner-by-region?expand=1#diff-c01c536742e880075c13cb24fefa1060R38). 
- [ ] On CODE (optional) note, if you want to test caching part on code you will need to temporarily update the cache purging logic as currently it only works on PROD, talk to Dotcom.

### CODE REVIEW

File: static/src/javascripts/lib/geolocation.js [hunk 1]
```diff
@@ -4,7 +4,6 @@ import config from 'lib/config';
 import { local as storage } from 'lib/storage';
 
 const storageKey = 'gu.geolocation';
-export type geolocationString = 'GB' | 'US' | 'AU';
```
Comment (tsop14) [inline static/src/javascripts/lib/geolocation.js]: Removing enum type as it is not all geolocation strings but limited to those relevant for reader revenue and so confusing. This type was only used in one test.

File: static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js [hunk 2]
```diff
@@ -37,6 +37,16 @@ export type CtaUrls = {
     supportUrl: string,
 };
 
+export type ReaderRevenueRegion = 'gb' | 'us' | 'au' | 'row';
+
+const getReaderRevenueRegion = (geolocation: string): ReaderRevenueRegion => {
+    const location = geolocation.toLowerCase();
```
Comment (joelochlann) [inline static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js]: is there any need to make them lowercase when they're always uppercase in dotcom?

**Review by joelochlann — APPROVED** [review]
What a thing to pull out of the bag on your second-to-last day! ⭐️ Thank you!

Have we put some JSON for each region in S3 that has the current last deploy time? So we don't have to redeploy each region immediately after merging

File: static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js (hunk 2 above)
Comment (joelochlann) [reply static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js]: as discussed in person - maybe full words are better to distinguish from genuine country codes? `'united-kingdom' | 'united-states' | 'australia' | 'rest-of-world' `

Comment (tsop14) [reply static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js]: This was to make them distinct from the geolcation (although they are determined by geolocation) to reduce chance they were accidentally used. I think your suggested making them clearly different by using words is a good one, especially as rest-of-world would not longer be confusing row 🚣

**Review by philmcmahon — APPROVED** [review]
Cache purge stuff looks good! Great last day PR 😃 

mic drop

**Comment by tsop14** [comment]
> Have we put some JSON for each region in S3 that has the current last deploy time? So we don't have to redeploy each region immediately after merging

If there is nothing in s3 it will return false for `hasBannerBeenRedeployedSinceClosed `. But you are right that a change was needed to prevent all users seeing banner! Have added in logic that will look for the old timestamp if there is no timestamp found for the region. This can be deleted eventually but means on merge the user's last known bannerClosedAt value is respected.

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 3]
```diff
@@ -33,17 +38,20 @@ type BannerDeployLog = {
...
         mode: 'cors',
     }).then((resp: BannerDeployLog) => resp && resp.time);
 
 const hasBannerBeenRedeployedSinceClosed = (
-    userLastClosedBannerAt: string
+    userLastClosedBannerAt: string,
+    region: ReaderRevenueRegion
 ): Promise<boolean> =>
-    getTimestampOfLastBannerDeploy()
+    getTimestampOfLastBannerDeployForLocation(region)
```
Comment (joelochlann) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: as discussed - even though the logic means we won't re-display the banner if there's nothing in S3, we will fire a LOT of Sentry errors - so let's put something in S3 before we merge

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js [hunk 4]
```diff
@@ -152,13 +160,23 @@ const selectSequentiallyFrom = (array: Array<string>): string =>
...
 const hideBanner = (banner: Message) => {
     banner.hide();
-
-    // Store timestamp in localStorage
-    userPrefs.set(lastClosedAtKey, new Date().toISOString());
+    const geolocation = geolocationGetSync();
+    const region: ReaderRevenueRegion = getReaderRevenueRegion(geolocation);
+
+    // Store timestamp for this geolocation banner in localStorage
+    const allEngagementBannerLastClosedAtTimestamps =
```
Comment (jranks123) [inline static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: might be nice to put a return type here

File: static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js (hunk 3 above)
Comment (tsop14) [reply static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js]: Yes! No sentry spam this christmas!

**Review by jranks123 — APPROVED** [review]

**Pull request merged**